flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSchemaUtils.java

```diff
@@ -68,6 +68,36 @@ public static TableSchema getPhysicalSchema(TableSchema tableSchema) {
         return builder.build();
     }
 
+    /**
+     * Return {@link TableSchema} which consists of all persisted columns. That means, the virtual
+     * computed columns and metadata columns are filterd out.
+     *
+     * <p>Readers(or writers) such as {@link TableSource} and {@link TableSink} should use this
+     * persisted schema to generate {@link TableSource#getProducedDataType()} and {@link
+     * TableSource#getTableSchema()} rather than using the raw TableSchema which may contains
+     * additional columns.
+     */
+    public static TableSchema getPersistedSchema(TableSchema tableSchema) {
```

Comment:
we'd better extract a common method for this and `getPhysicalSchema` since only one condition differs